identity/src/commonMain/kotlin/com/android/identity/nfc/NfcIsoTag.kt

```diff
@@ -0,0 +1,218 @@
+package com.android.identity.nfc
+
+import kotlinx.coroutines.delay
+import kotlinx.io.bytestring.ByteString
+import kotlin.time.Duration
+
+/**
+ * Class representing a ISO/IEC 14443-4 tag.
+ *
+ * This is an abstract super class intended for OS-specific code to implement the [transceive] method.
+ */
+abstract class NfcIsoTag {
+    /**
+     * Sends an APDU to the tag and waits for a response APDU.
+     *
+     * @param apdu the [CommandApdu] to send.
+     * @return the [ResponseApdu] which was received.
+     * @throws NfcTagLostException if the tag was lost.
+     */
+    abstract suspend fun transceive(apdu: CommandApdu): ResponseApdu
+
+    /**
+     * Selects an application according to ISO 7816-4 clause 11.2.2.
+     *
+     * @param applicationId the application to select, e.g. [Nfc.NDEF_APPLICATION_ID].
+     */
+    suspend fun selectApplication(applicationId: ByteString) {
+        // ISO 7816-4 clause 11.2.2
+        val response = transceive(
+            CommandApdu(
+                cla = 0,
+                ins = Nfc.INS_SELECT,
+                p1 = Nfc.INS_SELECT_P1_APPLICATION,
+                p2 = Nfc.INS_SELECT_P2_APPLICATION,
+                payload = applicationId,
+                le = 0
+            )
+        )
+        if (response.status != Nfc.RESPONSE_STATUS_SUCCESS) {
+            throw IllegalStateException("Error selecting application, status ${response.statusHexString}")
+        }
+    }
+
+    /**
+     * Selects a file according to ISO 7816-4 clause 11.2.2.
+     *
+     * @param fileId the identifier for the file to select, e.g. [Nfc.NDEF_CAPABILITY_CONTAINER_FILE_ID].
+     */
+    suspend fun selectFile(fileId: Int) {
+        //
+        val response = transceive(
+            CommandApdu(
+                cla = 0,
+                ins = Nfc.INS_SELECT,
+                p1 = Nfc.INS_SELECT_P1_FILE,
+                p2 = Nfc.INS_SELECT_P2_FILE,
+                payload = ByteString(encodeShort(fileId)),
+                le = 0
+            )
+        )
+        if (response.status != Nfc.RESPONSE_STATUS_SUCCESS) {
+            throw IllegalStateException("Error selecting file, status ${response.statusHexString}")
+        }
+    }
+
+    /**
+     * Reads binary data according to ISO 7816-4 clause 11.3.3.
+     *
+     * @param offset offset of where to read from.
+     * @param length amount of data to read, must be positive.
+     * @return the data which was read.
+     */
+    suspend fun readBinary(offset: Int, length: Int): ByteArray {
+        require(length > 0) { "Length must be positive" }
```

Comment:
Check that the offset fits in 2 bytes?
...Unless we're already adding that check to CommandApdu, in which case there's no need to check in both places.

Reply:
Done